src/dataflow/sink/avro_ocf.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::fs::File;
+
+use differential_dataflow::hashable::Hashable;
+use log::error;
+use timely::dataflow::channels::pact::Exchange;
+use timely::dataflow::operators::generic::Operator;
+use timely::dataflow::{Scope, Stream};
+
+use dataflow_types::{AvroOcfSinkConnector, Diff, Timestamp};
+use expr::GlobalId;
+use interchange::avro::{DiffPair, Encoder};
+use repr::{RelationDesc, Row};
+
+pub fn avro_ocf<G>(
+    stream: &Stream<G, (Row, Timestamp, Diff)>,
+    id: GlobalId,
+    connector: AvroOcfSinkConnector,
+    desc: RelationDesc,
+) where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let res = File::create(&connector.path);
+    let sink_file = match res {
+        Ok(f) => f,
+        Err(e) => {
+            error!("creating avro ocf file for sink failed: {}", e);
+            return;
+        }
+    };
+    let sink_hash = id.hashed();
+    let encoder = Encoder::new(desc);
+    let schema = encoder.writer_schema();
+    let mut avro_writer = avro::Writer::new(schema.clone(), sink_file);
```

Comment:
On second thought, this is going to be terribly broken for multi-worker clusters, right? Because they'll all create files that will stomp all over one another?

I think this needs to be guarded with:

```rust
let write_file = sink_hash % scope.peers() == scope.index();
let avro_writer = if write_file {
    match res {
        Ok(f) => {
            // ...
           Some(avrò::Writer::new(...))
        }
        Err(e) => {
            error!("creating avro ocf file for sink failed: {}", e);
            None
        }
    }
} else {
    None
};
```

and then inside the sink closure wrap everything in an `if let Some(avro_writer) = avro_writer`. I believe that fixes the divergence—we always create the sink, even if we can't create the file—and also ensures that at most one worker will create and write to the file.

Reply:
yep I forgot that create will truncate the file if it already exists :|

I looked at this https://doc.rust-lang.org/std/fs/struct.OpenOptions.html (I forget why I didn't look at this earlier I think I mistakenly thought that using it was nightly only), and I think we can do something a little bit nicer that avoids the synchronization on dataflow workers / relying on internal details:

1. In the post planning: lets create the file with create_new and fail if that fails
2. In the dataflow -- have every writer open a handle with append set to true (but not create or truncate) and log errors if that creation fails, and then have every writer create the sink / wrap writes by checking if the writer is none as you suggested

I'll write this up and we can see if that works better